design/open/branch_checks.md

```diff
@@ -0,0 +1,171 @@
+This puts a pretty short time window for hooks to complete in.
+
+Distributed processing frameworks such as Apache Spark typically have a long startup time and might take minutes (or even hours) to complete.
+
+The main goal of this proposal is to allow such validations to occur with a very long timeout (configurable, but could be days).
+
+## Non-Goals
+
+Out of scope for this proposal is any collaboration features that are typically associated with PRs (that provide similar functionality to the one being proposed).
+
+## High Level design
+
+This proposal adds a `Check` primitive to lakeFS branches. a Check is similar to a hook: it calls out to some external component with information about a modification or event that occurred in the system.
+
+The user flow for checks is the following:
+
+1. configure checks
+1. ensure all *mandatory* checks have passed for the given branch's HEAD
+1. merge said branch into a protected one.
+
+Merging into any protected branch requires all mandatory checks to have completed successfully for the current HEAD:
+
+```mermaid
+sequenceDiagram
+    Actor user1
+    user1->>lakeFS: commit1(branch_a)
+    user1->>lakeFS: commit2(branch_a)
+    user1->>lakeFS: commitN(branch_a)
+    user1->>lakeFS: check(branch_a) // will be dereferenced
+    loop For every configured check
+      lakeFS->>+External Executor: execute(event details, callback URL)
+      External Executor->>-lakeFS: callback(status)
+      lakeFS->>lakeFS: update check status (SUCCESS)
+    end
+    user1->>lakeFS: merge(branch_a, branch_b_protected)
+    lakeFS-->>user1: success!
+```
+
+## Implementation
+
+#### Check configuration
+
+Checks are configured under `_lakefs_actions/` within a lakeFS repository's root. Any YAML file within the directory can contain checks, similarly to hooks.
+Here's what a check would look like:
+
+```yaml
+name: Ensure Event counts are within bounds
+checks:
+  - id: great_expectations_validate_events
+    type: webhook
+    properties:
+      url: "http://<host:port>/checks/great_expectations"
+      query_params:
+        condition: expect_table_row_count_to_be_between(min_value=2000, max_value=5000)
+      headers:
+        secret_header: "{{ ENV.MY_SECRET }}"
+```
+
+#### Check event syntax
+
+```json
+{
+  "repository_id": "example-repo",
+  "branch_id": "branch_a",
+  "source_ref": "0307e79ab5afef2b0726a8c13896f716033b1acd8454e22210bd43a2f72a94a9",
+  "check_id": "great_expectations_validate_events",
+  "storage_namespace": "s3://path/to/storage/",
+  "callback_token": "abc123",
+  "output_url": "presigned URL or another HTTP POST-able URL"
+}
+```
+
+#### Check triggering
+
+Checks are triggered when the `check(repository_id, ref_id)` endpoint is called in the lakeFS API.
+When running check() with a branch or tag, it will be dereferenced: checks are valid for a given commit ID.
+
+
+#### Check execution
```

Comment:
If I understand correctly - checks are run under "branch_a" code base - while status check is done from "branch_b" context.
Some questions:
1. If I want to remove a check from "branch_b" how can I do it and still be able to provided a successful status for it.
2. What is stopping an "uneducated" user (or malicious?) to modify checks to return "always true" and bypass the checks mechanism?

Reply:
Good questions :)
1. pre-merge, the destination branch can validate that the required check ids have passed. if source branch didn't run them, they can merge destination into source (or rebase in the future) to make sure they have the correct checks.
2. use a policy that doesn't allow uneducated users (or say, only allows a specific power group) to write to `_lakefs_actions/` - this is the same answer for hooks.

Reply:
I seem to be confused a bit on no.1
Let's say for example that I want to remove a required check on "branch_b". I create a branch called "branch_a" which removes a required check from "branch_b" and commit the changes. Now I run the check API against "branch_a" which creates a list of check results on the given head commit ref that is missing the specific check. Now when I want to merge "branch_a" to "branch_b" I have a problem since there is no status for a required check in "branch_b" for that given reference

Reply:
As long as mandatory check validation happens as a pre-merge hook (either directly or behind the scenes), hooks should run using a the merged state of the `_lakefs_actions/` directory. 

I'm not sure this is currently the case - but if not, it means that the problem is how we can update pre-merge hooks on protected branches. Checks should be mostly orthogonal to this..